### PR TITLE
ECS: Add logger

### PR DESCRIPTION
**Pull request by Saverio976**
MINOR

* **Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [x] Norm of the code has been respected

* **In what state is your pull request?**
- [x] Ready to merge / Waiting a reviwer to see my work.
- [ ] Work In Progress (WIP) / My work is not finish but i want daily reviews. (Draft)
- [ ] CI Review / I want to see what the CI thinks of my work.
- [ ] Need feedback / Just to have feedback on what i produced. (May not be merged)

* **What kind of change does this PR introduce?** (You can choose multiple)
- [ ] Bug fix
- [ ] Feature request
- [x] New / Updated documentation
- [ ] Testing CI ( Make the pull request in draft mode)

* **What is the current behavior?** (link an issue based on the kind of change this pr introduce)

* **What is the new behavior (if this is a feature change)?**

* **Other information**:

### CODE REVIEW

**Review by guillaumeAbel — APPROVED** [review]

File: src/ECS/Logger.cpp [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+** EPITECH PROJECT, 2023
+** Raytracer
+** File description:
+** Logger.cpp
+*/
+
+#include "Logger.hpp"
+#include <ctime>
+#include <iostream>
+#include <map>
```
Comment (kw4ntiq) [inline src/ECS/Logger.cpp]: Already include in logger.hpp

File: src/ECS/Logger.cpp [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+#include <string>
```
Comment (kw4ntiq) [inline src/ECS/Logger.cpp]: Already include in logger.hpp

File: src/ECS/Logger.hpp [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+#define _LOG_ERROR_VALUE 1
+#define _LOG_WARN_VALUE  2
+#define _LOG_INFO_VALUE  3
+#define _LOG_DEBUG_VALUE 4
+#define _LOG_TRACE_VALUE 5
+#define _LOG_RESET_COLOR 6
+
+namespace Logger {
+    /**
+     * @brief Logger singleton
```
Comment (kw4ntiq) [inline src/ECS/Logger.hpp]: Singleton ?

File: src/ECS/Logger.cpp [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+        static std::map<int, std::string> colors =
+            {
+                {_LOG_FATAL_VALUE, ""},
+                {_LOG_ERROR_VALUE, ""},
+                {_LOG_WARN_VALUE,  ""},
+                {_LOG_INFO_VALUE,  ""},
+                {_LOG_DEBUG_VALUE, ""},
+                {_LOG_TRACE_VALUE, ""},
+                {_LOG_RESET_COLOR, ""},
+        }
```
Comment (kw4ntiq) [inline src/ECS/Logger.cpp]:
```suggestion
        };
```

File: src/ECS/Logger.cpp [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+        time.erase(time.find_last_of("\n"));
+        mes = time + " [" + level + "] " + message;
+        std::cerr << colors[levelT] << mes << colors[_LOG_RESET_COLOR]
+                  << std::endl;
+        if (it != _callbacks.end()) {
+            for (auto &it1 : it->second) {
+                it1.second(mes);
+            }
+        }
+    }
```
Comment (kw4ntiq) [inline src/ECS/Logger.cpp]: Add preprocessor directives (#if WIN32 etc), it won't compile on windows (std::localtime is not a thread safe function + print logic for colors will be completely different here).

File: src/ECS/Logger.cpp [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+#endif
+    }
+
+    void Logger::print(
+        int levelT,
+        const std::string &level,
+        const std::string &message)
+    {
+#ifdef __linux__
+        static std::map<int, std::string> colors = {
```
Comment (kw4ntiq) [inline src/ECS/Logger.cpp]: Is it important to have this map in order?

File: src/ECS/Logger.cpp [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+            {_LOG_FATAL_VALUE, "\033[31m"},
+            {_LOG_ERROR_VALUE, "\033[33m"},
+            {_LOG_WARN_VALUE,  "\033[34m"},
+            {_LOG_INFO_VALUE,  "\033[32m"},
+            {_LOG_DEBUG_VALUE, "\033[38m"},
+            {_LOG_TRACE_VALUE, "\033[30m"},
+            {_LOG_RESET_COLOR, "\033[0m" },
+        };
+#elif __APPLE__
+        static std::map<int, std::string> colors = {
```
Comment (kw4ntiq) [inline src/ECS/Logger.cpp:117]: same

File: src/ECS/Logger.hpp [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+            void unsubscribeCallback(int type, const std::string &name);
+
+        private:
+            void print(
+                int levelT,
+                const std::string &level,
+                const std::string &message);
+            std::map<
+                int,
+                std::map<std::string, std::function<void(const std::string &)>>>
```
Comment (kw4ntiq) [inline src/ECS/Logger.hpp:150]: Is it important to have this map in order?

**Review by kw4ntiq — CHANGES_REQUESTED** [review]

File: src/ECS/Logger.cpp (hunk 1 above)
Comment (Saverio976) [reply src/ECS/Logger.cpp]: vote

File: src/ECS/Logger.cpp (hunk 2 above)
Comment (Saverio976) [reply src/ECS/Logger.cpp]: vote

File: src/ECS/Logger.cpp (hunk 6 above)
Comment (Saverio976) [reply src/ECS/Logger.cpp]: opti to use this

**Review by kw4ntiq — APPROVED** [review]

**Pull request merged**